### ISSUE TITLE
20 times faster installation times

### ISSUE THREAD

**Issue by jantimon** [google/zx#669]
last week I created an experiment for `zx`, which might be interesting for you

my experiment is an optimized node bundle of `zx` created using the Vercel `ncc` compiler, offering significant size reduction from 13mb (incl dependencies) to 450kb (with 0 dependencies).

`du -sh node_modules`
![shot-6uNcUrYK](https://github.com/google/zx/assets/4113649/5d45d2d1-1276-4aa3-aab2-94acaad9e812)
![shot-qkpNxRzE](https://github.com/google/zx/assets/4113649/ea57bdb1-2521-4ef9-b92c-c9182ccfd45a)

you can find it on GitHub [here](https://github.com/jantimon/zx-packed) and on npm [here](https://www.npmjs.com/package/zx-packed)

[![shot-zVD6hOxM](https://github.com/google/zx/assets/4113649/bd9bdf77-3a8d-436b-a6f1-bdd63bdd16e1)](https://www.npmjs.com/package/zx-packed)

a lot of users of `zx` will probably prefer to use a non bundle package but still I would love to know what you think :)

**Comment by antonmedv** [comment]
Interesting idea. I will think about it.

**Comment by njfamirm** [comment]
If I understand correctly, you want to build the codes before publishing in npm, right?
What exactly is done? @jantimon

**Comment by jantimon** [comment]
@njfamirm from the **ncc** readme on github https://github.com/vercel/ncc

> Simple CLI for compiling a Node.js module into a single file,
> together with all its dependencies, gcc-style.
> 
> ## Motivation
> 
> - Publish minimal packages to npm
> - Only ship relevant app code to serverless environments
> - Don't waste time configuring bundlers
> - Generally faster bootup time and less I/O overhead
> - Compiled language-like experience (e.g.: `go`)
> 
> ## Design goals
> 
> - Zero configuration
> - TypeScript built-in
> - Only supports Node.js programs as input / output
> - Support all Node.js patterns and npm modules

In my experiment `ncc` was used as it bundles everything into a single js files and removes all unnecessary parts with treeshaking.

It also minifies the code:
https://unpkg.com/browse/zx-packed@0.0.6/dist/index.mjs

The very same would probably be possible with esbuild (see [https://esbuild.github.io/getting-started/#bundling-for-node](https://esbuild.github.io/getting-started/#bundling-for-node)) but might require a little bit more configuration.

What do you mean by "What exactly is done?"?
The experiment works: 

```bash
npm i -g zx-packed
zx your-script.js
```

it also works with a shebang:

```bash
#!/usr/bin/env zx
// ...
```

it also works if you import it:

```
import "zx-packed";
// ...
```

**Comment by njfamirm** [comment]
Thank you for your complete explanation ❤️ @jantimon 

In my opinion and according to the searches I did, `esbuld` is a better option than `ncc` and has extremely high speed!
Configuring it is not that much and I can do it.

**Comment by njfamirm** [comment]
The side question I have is where do you get such a screenshot ?

**Comment by antonmedv** [comment]
Another thing is zx/glovals endpoint.

**Comment by jantimon** [comment]
Right now `globals` is not explicitly listed as export in the package.json of my experimental package.

However that could be easily changed and it is already possible with a less pretty import path:

![shot-gyNyc4JV](https://github.com/google/zx/assets/4113649/a66e4658-7776-4b87-be88-dad471306373)

**Comment by antonmedv** [comment]
Looks like globals just import index:
<img width="795" alt="image" src="https://github.com/google/zx/assets/141232/3c80e4da-e2cd-486a-887b-436a06695597">
Will be cool to have same logic as now.

**Comment by jantimon** [comment]
@antonmedv - I just released a new version: 0.0.8

both ways are now fully supported - named and global usage:

![demo of global and named imports](https://github.com/google/zx/assets/4113649/5fd487d8-3a39-411e-821d-93aa5c7e4b1a)

@antonmedv I switched to esbuild which solved some ncc issue but created other issues :)

I had to patch `zx/src/cli.js` as esbuild does not allow top level awaits

A fresh version can be found 

 - https://github.com/jantimon/zx-packed
 - https://unpkg.com/browse/zx-packed@0.0.8/dist/
 - https://www.npmjs.com/package/zx-packed/v/0.0.8

Using esbuild further reduced the filesize from 516K to 448K.

**Comment by antongolub** [comment]
Seems covered by https://github.com/google/zx/pull/722
